python_modules/libraries/dagster-dg/dagster_dg/cli/scaffold.py

```diff
@@ -115,12 +115,94 @@ def _define_commands(self, cli_context: click.Context) -> None:
         dg_context = DgContext.for_defined_registry_environment(Path.cwd(), config)
 
         registry = RemotePluginRegistry.from_dg_context(dg_context)
+
+        # Keys where the actual class name is not shared with any other key will use the class name
+        # as a command alias.
+        keys_by_name: dict[str, set[PluginObjectKey]] = {}
+        for key in registry.keys():
+            keys_by_name.setdefault(key.name, set()).add(key)
+
         for key, component_type in registry.items():
-            command = _create_scaffold_defs_subcommand(key, component_type)
-            self.add_command(command)
+            self._create_subcommand(
+                key, component_type, use_typename_alias=len(keys_by_name[key.name]) == 1
+            )
 
         self._commands_defined = True
 
+    def _create_subcommand(
+        self, key: PluginObjectKey, obj: PluginObjectSnap, use_typename_alias: bool
+    ) -> None:
+        # We need to "reset" the help option names to the default ones because we inherit the parent
+        # value of context settings from the parent group, which has been customized.
+        @self.command(
+            cls=ScaffoldDefsSubCommand,
+            name=key.to_typename(),
+            context_settings={"help_option_names": ["-h", "--help"]},
+            aliases=[key.name] if use_typename_alias else None,
+        )
+        @click.argument("instance_name", type=str)
```

Comment:
are we concerned about scaling here? I'm just thinking about a world where there are 10k components in a large python environment (not farfetched at scale IMO) and this gets slow.

Reply:
I also realize it predates this PR

Reply:
Yes, I think the approach here will have to be changed eventually if we are in a world where we're generating huge numbers of subcommands.